test/extensions/filters/http/ext_authz/config_test.cc

```diff
@@ -3,22 +3,177 @@
+      // i must be explicitly captured by value.
+      workers_.emplace_back(api_->threadFactory().createThread(
+          [&, i]() { dispatchers_[i]->run(Event::Dispatcher::RunType::RunUntilExit); }));
+    }
+    // Post callback to fire all threads at the same time.
+    for (Event::DispatcherPtr& dispatcher : dispatchers_) {
+      dispatcher->post([&]() {
+        waitForAllWorkersToFire();
+        workerFunc();
+      });
+    }
+    // Post exit signals and wait for thread to end.
+    for (Event::DispatcherPtr& dispatcher : dispatchers_) {
+      dispatcher->post([&dispatcher]() { dispatcher->exit(); });
+    }
+    for (Thread::ThreadPtr& worker : workers_) {
+      worker->join();
+    }
+  }
+
+  void waitForAllWorkersToFire() {
+    if (shouldFire()) {
+      // The last thread spawned should notify all to start.
+      workers_should_fire_.Notify();
+    } else {
+      // Wait for notification if the current thread is not the last one spawned.
+      workers_should_fire_.WaitForNotification();
+    }
+  }
+
+  bool shouldFire() {
+    Thread::LockGuard lock(mutex_);
+    active_threads_cnt_++;
+    return active_threads_cnt_ == num_threads_;
+  }
+
+  size_t num_threads_;
+  absl::Notification workers_should_fire_;
+  Thread::MutexBasicLockable mutex_;
+  size_t active_threads_cnt_{};
+  ThreadLocal::InstanceImpl tls_;
+  Api::ApiPtr api_;
+
+  std::vector<Event::DispatcherPtr> dispatchers_;
+  std::vector<Thread::ThreadPtr> workers_;
+};
+
+class ExtAuthzFilterTest : public MultiThreadTest, public testing::Test {
+public:
+  ExtAuthzFilterTest() : MultiThreadTest(10), stat_names_(symbol_table_) {}
+
+  void initialize(const std::string& ext_authz_config_yaml) {
+    initFilterFactory(ext_authz_config_yaml);
+    initAddress();
+  }
+
+  void initFilterFactory(const std::string& ext_authz_config_yaml) {
+    EXPECT_CALL(context_, getServerFactoryContext())
+        .WillRepeatedly(testing::ReturnRef(server_context_));
+    ON_CALL(context_.cluster_manager_.async_client_manager_, getOrCreateRawAsyncClient(_, _, _, _))
+        .WillByDefault(Invoke([&](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool,
+                                  Grpc::CacheOption) {
+          return std::make_unique<NiceMock<Grpc::MockAsyncClient>>();
+        }));
+    ExtAuthzFilterConfig factory;
+    ProtobufTypes::MessagePtr proto_config = factory.createEmptyConfigProto();
+    TestUtility::loadFromYaml(ext_authz_config_yaml, *proto_config);
+    filter_factory_ = factory.createFilterFactoryFromProto(*proto_config, "stats", context_);
+  }
+
+  void initAddress() {
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+    connection_.stream_info_.downstream_address_provider_->setRemoteAddress(addr_);
+    connection_.stream_info_.downstream_address_provider_->setLocalAddress(addr_);
+  }
+
+  void testFilter() {
+    Http::MockFilterChainFactoryCallbacks filter_callbacks;
```

Comment:
can some of the helper code in this test-class be factored out and shared with other test-classes in this file?